lib/server.js

```diff
@@ -373,6 +323,110 @@ WalletService.prototype._notify = function(type, data, opts, cb) {
 };
 
 
+WalletService.prototype._addCopayerToWallet = function(wallet, opts, cb) {
+  var self = this;
+
+  if (wallet.copayers.length == wallet.n) return cb(Errors.WALLET_FULL);
+
+  var copayer = Model.Copayer.create({
+    name: opts.name,
+    copayerIndex: wallet.copayers.length,
+    xPubKey: opts.xPubKey,
+    requestPubKey: opts.requestPubKey,
+    signature: opts.copayerSignature,
+  });
+
+  self.storage.fetchCopayerLookup(copayer.id, function(err, res) {
+    if (err) return cb(err);
+    if (res) return cb(Errors.COPAYER_REGISTERED);
+
+    wallet.addCopayer(copayer);
+    self.storage.storeWalletAndUpdateCopayersLookup(wallet, function(err) {
+      if (err) return cb(err);
+
+      async.series([
+
+        function(next) {
+          self._notify('NewCopayer', {
+            walletId: opts.walletId,
+            copayerId: copayer.id,
+            copayerName: copayer.name,
+          }, next);
+        },
+        function(next) {
+          if (wallet.isComplete() && wallet.isShared()) {
+            self._notify('WalletComplete', {
+              walletId: opts.walletId,
+            }, {
+              isGlobal: true
+            }, next);
+          } else {
+            next();
+          }
+        },
+      ], function() {
+        return cb(null, {
+          copayerId: copayer.id,
+          wallet: wallet
+        });
+      });
+    });
+  });
+};
+
+
+WalletService.prototype._addKeyToCopayer = function(wallet, copayer, opts, cb) {
+  var self = this;
+  wallet.addCopayerRequestKey(copayer.copayerId, opts.requestPubKey, opts.signature, opts.restrictions);
+  self.storage.storeWalletAndUpdateCopayersLookup(wallet, function(err) {
+    if (err) return cb(err);
+
+    return cb(null, {
+      copayerId: copayer.id,
+      wallet: wallet
+    });
+  });
+};
+
+/**
+ * Adds access to a given copayer
+ *
+ * @param {Object} opts
```

Comment:
Would it make sense to accept a text field as an identifier for this particular access key? This way we could add access keys for agents.